doc/nvgre_tunnel/nvgre_tunnel.md

```diff
@@ -0,0 +1,437 @@
+
+From the architecture point of view the new orchagent daemon will be added to cover the NVGRE functionality. New daemon should handle the configuration taken from config DB, and call an appropriate SAI API, which will create tunnel and tunnel maps.
+
+### Requirements
+
+This section describes the SONiC requirements for NVGRE feature.
+
+At a high level the following should be supported:
+
+- User should be able to create NVGRE tunnel
+- User should be able to create VLAN to VSID mapper entries for the NVGRE tunnel
+- Both VLAN and Bridge to VSID mappers should be supported by the NVGRE tunnel
+- Only the decapsulation mappers supported
+- YANG model should be created in order to auto-generate CLI by using the [SONiC CLI Auto-generation tool](https://github.com/Azure/SONiC/blob/master/doc/cli_auto_generation/cli_auto_generation.md).
+- CLI for NVGRE tunnel
+
+Counters for NVGRE Tunnel are out of scope of this design document.
+
+### Architecture Design 
+
+The new tables will be added to Config DB. Unless otherwise stated, the attributes are mandatory.
+
+#### NVGRE configDB table
+
+```
+NVGRE_TUNNEL|{{tunnel_name}} 
+    "src_ip": {{ip_address}} 
+
+NVGRE_TUNNEL_MAP|{{tunnel_name}}|{{tunnel_map}}
+    "vsid": {{vsid_id}}
+    "vlan": {{vlan_id}}
+```
+
+#### ConfigDB schemas
+
+```
+; Defines schema for NVGRE Tunnel configuration attributes
+key                                   = NVGRE_TUNNEL:name             ; NVGRE tunnel configuration
+; field                               = value
+SRC_IP                                = ipv4                          ; IPv4 source address
+
+;value annotations
+ipv4          = dec-octet "." dec-octet "." dec-octet "." dec-octet
+dec-octet     = DIGIT                     ; 0-9  
+                  / %x31-39 DIGIT         ; 10-99  
+                  / "1" 2DIGIT            ; 100-199  
+                  / "2" %x30-34 DIGIT     ; 200-249
+```
+
+```
+; Defines schema for NVGRE Tunnel map configuration attributes
+key                                   = NVGRE_TUNNEL:tunnel_name:name ; NVGRE tunnel configuration
+; field                               = value
+VSID                                  = DIGITS                        ; 1 to 16 million values
+VLAN                                  = 1\*4DIGIT                     ; 1 to 4094 Vlan id
+```
+
+#### Orchestration agent
+
+The following orchestration agents will be added or modified. The flow diagrams are captured in a later section.
+
+##### Figure 1. Orchestration agents
+
+<p align=center>
+<img src="images/nvgre_orch.svg" alt="Figure 1. Orchestration agents">
+</p>
+
+#### NvgreOrch
+
+`nvgreorch` - it is an orchestration agent that handles the configuration requests directly from ConfigDB. The `nvgreorch` is responsible for creating the tunnel and attaches encap and decap mappers. Both VLAN and Bridge tunnel maps are created for NVGRE tunnel by default, so it would be possible to apply both VLAN/VSID and Bridge/VSID map entries.
```

Comment:
No plan to have nvgremgr to handle the config-DB events to program the kernel and update the APP_DB for nvgreorch to handle and update the ASIC_DB? please mention that the scope of this HLD is only for NVGRE tunnel termination.

Reply:
No, we don't plan to have the `nvgremgr` because there is nothing to configure in the kernel for NVGRE Tunnel.
In this case, we can omit the configuration to `APP_DB` and pass the configuration directly to `CONFIG_DB`.

Also, why do we need to mention that the scope of this HLD is only for NVGRE tunnel termination?

Reply:
I'm referring to your below comment on the decap support, I want you to mention the scope in the HLD.

@tzack000 currently we add implementation only for the decap flow and it is L2 over L3 tunnel, so it should not be referenced by route objects. So at this point, you will be able to configure tunnel object and mapper entries for decap. This covers exactly the same use case as we have for L2 VXLAN, for example, please take a look at this test - test_vxlan_decap.py

Reply:
Sure, fixed